### PR TITLE
Implemented graceful shutdown

### PR DESCRIPTION
**Pull request by dk-neon**
Also updated workflow in order to prevent message loss in filter.

### CODE REVIEW

File: v1-simple/src/db.rs [hunk 1]
```diff
@@ -76,223 +83,128 @@ pub async fn create_db_pool(config: Arc<AppConfig>) -> Result<Arc<Pool>> {
...
+    let mut shutdown_stream = signal(SignalKind::terminate()).unwrap();
+
+    let client = loop {
+        select! {
+            _ = shutdown_stream.recv() => return,
+            db_pool_result = db_pool.get() => match db_pool_result {
+                Ok(client) => break Arc::new(client),
+                Err(err) => {
+                    error!("Failed to get a client from the pool, error: {err:?}");
+                    tokio::time::sleep(Duration::from_millis(50)).await;
```
Comment (Deniskore) [inline v1-simple/src/db.rs:185]: Let's switch to interval here.

File: v1-simple/src/offset_manager.rs [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+            return 0;
+        }
+        index
+    }
+}
+
+pub struct OffsetManager {
+    topic: String,
+    consumer: Arc<StreamConsumer<ContextWithStats>>,
+    partitions: HashMap<i32, PartitionOffsetManager>,
```
Comment (Deniskore) [inline v1-simple/src/offset_manager.rs:171]: Replace with https://docs.rs/ahash/latest/ahash/struct.AHashMap.html please.

File: v1-simple/src/db.rs [hunk 3]
```diff
@@ -76,223 +83,128 @@ pub async fn create_db_pool(config: Arc<AppConfig>) -> Result<Arc<Pool>> {
...
-    let (account_tx, account_rx) = account_queue;
-    let (slot_tx, slot_rx) = slot_queue;
-    let (transaction_tx, transaction_rx) = transaction_queue;
-    let (block_tx, block_rx) = block_queue;
+    process_msg_async: impl Fn(Arc<Client>, Arc<M>) -> F,
+) where
+    F: Future<Output = Result<u64>>,
+{
+    let (message, offset) = queue_msg;
+    let mut shutdown_stream = signal(SignalKind::terminate()).unwrap();
```
Comment (Deniskore) [inline v1-simple/src/db.rs:176]: I suggest handling this unwrap.

File: v1-simple/src/consumer.rs [hunk 4]
```diff
@@ -141,27 +146,48 @@ pub async fn consumer<T, S>(
...
+    consumer: Arc<StreamConsumer<ContextWithStats>>,
+    filter_config: Arc<RwLock<FilterConfig>>,
+    topic: String,
+    filter_tx: Sender<QueueMsg<S>>,
+    ctx_stats: ContextWithStats,
+) where
+    T: for<'a> Deserialize<'a> + Send + 'static + GetMessageType + GetEvent,
+    S: From<T> + Send + 'static,
+{
+    let mut shutdown_stream = signal(SignalKind::terminate()).unwrap();
```
Comment (Deniskore) [inline v1-simple/src/consumer.rs:166]: I suggest handling this unwrap with a good error message.

**Review by Deniskore — CHANGES_REQUESTED** [review]

File: v1-simple/src/offset_manager.rs (hunk 2 above)
Comment (dk-neon) [reply v1-simple/src/offset_manager.rs:171]: Done.

File: v1-simple/src/db.rs (hunk 3 above)
Comment (Deniskore) [inline v1-simple/src/db.rs:176]: @dk-neon Need to think about how not to create a signal handler on every message.

Comment (dk-neon) [reply v1-simple/src/db.rs:176]: Done.

File: v1-simple/src/consumer.rs (hunk 4 above)
Comment (dk-neon) [reply v1-simple/src/consumer.rs:166]: Done.

**Comment by Deniskore** [comment]
@dk-neon Please add new parameter to README.md.

**Review by Deniskore — APPROVED** [review]

**Pull request merged**